herd/AArch64Sem.ml

```diff
@@ -1333,6 +1325,28 @@ module Make
         let ops = List.map op (Misc.interval 0 (neon_nelem (List.hd rlist))) in
         List.fold_right (>>::) ops (M.unitT [[()]])
 
+      (* Data cache operations *)
+      let dc_loc op a ii =
+        let mk_act loc = Act.DC (op,Some loc) in
+        let loc = A.Location_global a in
+        M.mk_singleton_es (mk_act loc) ii
+
+      let do_dc op rd ii =
+        if AArch64Base.DC.sw op then
+          M.mk_singleton_es (Act.DC (op, None)) ii >>! B.Next
+        else begin
+            (* TODO: The size for DC should be a cache line *)
+            let mop _ac a = dc_loc op a ii in
+            lift_memop Dir.R
+              (fun ac ma _mv -> (* value fake here *)
+                if Access.is_physical ac then
+                  M.bind_ctrldata ma (mop ac)
```

Comment:
I wanted to merge this PR. But now I am not so sure about the control dependency. Can you please comment?

Reply:
This is very similar to what we do for regular load/store instructions https://github.com/herd/herdtools7/blob/cc5ae26620ff3191c215b10c932f695d3de1c7c8/herd/AArch64Sem.ml#L890. If I understand correctly, this has to do with the branching event we add between the `Rpte` event and the corresponding memory read/write event. DC instructions similarly to loads and stores may raise a fault if the PTE is invalid.

Reply:
I see that DC is assimilated to a read effect. This has some importance for the hardware update of the access and dirty bits. I'll explore this by some experiment before merging. Notice that documentation states that `DC CVAU` checks for read permission. Hence the `Dir.R` argument to `liftmem_op` looks appropriate.

Reply:
The clean or clean and invalidate DC variants (DC CVAU, DC CVAC, DC CIVAC) all require read permissions. There is one exception which is the invalidate DC IVAC instruction which requires write permissions. I need to check though if it sets the dirty bit.

Reply:
I have just checked on hardware that `DC CVAU` behaves as a load as regards hardware flag management...  I can check `DC IVAC`.

Reply:
As far as I have tested, `DC IVAC` require write permission. However if hardware management of the dirty bit is enabled, it no longer fails  but does *not* update the dirty bit...

Reply:
I think this version addresses the problem with DC IVAC, by raising a fault if it doesn't have write permissions but avoiding to update the DB if it's not set.

Reply:
You think right...